### ISSUE TITLE
Twitter/OAuth Provider AuthCredentialCallback in never fired.

### ISSUE THREAD

**Issue by Sagar-AppLogics** [firebase/firebase-ios-sdk#4414]
Environment:

- Xcode version: 11.2.1
- Firebase SDK version: 6.13.0
- Firebase Component: Auth
- Component version: 6.4.0

Problem:
Twitter Login using OAuthProvider is not working. The sign in to twitter screen is not even shown.

Setup twitter authentication using the following guide:
https://firebase.google.com/docs/auth/ios/twitter-login?authuser=0

Call the following as per step 5. "Authenticate with Firebase using the OAuth provider object", in my implementation (below) the line `print("NEVER CALLED...")` as suggested is never fired.
```
let provider = OAuthProvider(providerID: "twitter.com")
provider.getCredentialWith(nil) { (credential, error) in
    print("NEVER CALLED...")
    if let error = error {
        completion(.failure(error))
    } else {
        guard let credential = credential else {
            // TODO: Raise Error
            return
        }
        // TODO: Process credentials
    }
}
```

To summarise, when calling the function:
`getCredentialWith(_ UIDelegate: AuthUIDelegate?, completion: AuthCredentialCallback? = nil)`
the callback `AuthCredentialCallback` is never called.

Also tracing in to the file FirebaseAuth/FIROAuthProvider.m at line 144 `strongSelf` is `nil` and falls through

Breakpoint at line: 144
Snippet start line: 144:
```
[strongSelf getHeadFulLiteURLWithEventID:eventID
                         sessionID:sessionID
                        completion:^(NSURL *_Nullable headfulLiteURL,
                                     NSError *_Nullable error) {
    ...
}];
```
Snippet end line: 181

**Comment by rizafran** [comment]
Thanks for reporting @Sagar-AppLogics. I'll try to replicate this on my end and I'll let you know with my updates.

**Comment by rizafran** [comment]
@Sagar-AppLogics By using the [Auth quickstart](https://github.com/firebase/quickstart-ios/tree/master/authentication), the getCredentialWith() was called successfully. But when I changed the code to the one you have provided, I was able to replicate the issue. I noticed that when the OAuthProvider is declared at the top with `var provider: OauthProvider?` and did the following, the AuthCredentialCallback will be called:
```
self.provider = OAuthProvider(providerID: "twitter.com")
self.provider?.getCredentialWith(nil) {credential, error in
 print("NEVER CALLED...")
    if let error = error {
        completion(.failure(error))
    } else {
        guard let credential = credential else {
            // TODO: Raise Error
            return
        }
        // TODO: Process credentials
    }
}
```

Given this, I'll consult internally to check what's causing the issue. For now, you may try the approach I've provided above as a workaround.

**Comment by Sagar-AppLogics** [comment]
Thanks for the quick turn around on this one @rizafran I will try this out now and report back here.

**Comment by Sagar-AppLogics** [comment]
@rizafran Thanks again for the quick turn around on this, this indeed did solve the issue.
Also checked the and as long as its declared at a class or global level it seems to work, even as a constant - for example the following code seems to work without issue:

```
let provider = OAuthProvider(providerID: "twitter.com")

func SignIn() {
    provider.getCredentialWith(nil) { (credential, error) in
        print("NEVER CALLED...")
        if let error = error {
            completion(.failure(error))
        } else {
            guard let credential = credential else {
                // TODO: Raise Error
                return
            }
            // TODO: Process credentials
        }
    }
}
```

**Comment by Sagar-AppLogics** [comment]
@rizafran after runnings some more complete tests I realised this was only partially solved so I edited my previous response.

**Comment by rizafran** [comment]
Hi @Sagar-AppLogics. Thanks for letting me know. I'm glad to hear that you are now redirecting through the login process successfully. As for the other issue, I'll check your new post. :)

**Comment by rizafran** [comment]
Hi @Sagar-AppLogics, upon discussing with the engineer, it seems that the "provider" is released when the callback is actually called. It's better to avoid declaring it as a local variable.

I'm going to close this issue since it has been fixed. If you have other questions/clarifications, feel free to open this thread.
